clang/lib/CIR/CodeGen/CIRGenException.cpp

```diff
@@ -382,5 +410,279 @@ void CIRGenFunction::exitCXXTryStmt(const CXXTryStmt &s, bool isFnTryBlock) {
+  case EHScope::Terminate:
+    cgm.errorNYI("emitLandingPad: terminate");
+    return {};
+
+  case EHScope::Catch:
+  case EHScope::Cleanup:
+  case EHScope::Filter:
+    // CIR does not cache landing pads.
+    break;
+  }
+
+  // If there's an existing TryOp, it means we got a `cir.try` scope
+  // that leads to this "landing pad" creation site. Otherwise, exceptions
+  // are enabled but a throwing function is called anyways (common pattern
+  // with function local static initializers).
+  mlir::ArrayAttr handlerTypesAttr = tryOp.getHandlerTypesAttr();
+  if (!handlerTypesAttr || handlerTypesAttr.empty()) {
+    // Accumulate all the handlers in scope.
+    bool hasCatchAll = false;
+    llvm::SmallVector<mlir::Attribute, 4> handlerAttrs;
+    for (EHScopeStack::iterator i = ehStack.begin(), e = ehStack.end(); i != e;
+         ++i) {
+      switch (i->getKind()) {
+      case EHScope::Cleanup: {
+        cgm.errorNYI("emitLandingPad: Cleanup");
+        return {};
+      }
+
+      case EHScope::Filter: {
+        cgm.errorNYI("emitLandingPad: Filter");
+        return {};
+      }
+
+      case EHScope::Terminate: {
+        cgm.errorNYI("emitLandingPad: Terminate");
+        return {};
+      }
+
+      case EHScope::Catch:
+        break;
+      }
+
+      EHCatchScope &catchScope = cast<EHCatchScope>(*i);
+      for (unsigned handlerIdx = 0, he = catchScope.getNumHandlers();
+           handlerIdx != he; ++handlerIdx) {
+        EHCatchScope::Handler handler = catchScope.getHandler(handlerIdx);
+        assert(handler.type.flags == 0 &&
+               "landingpads do not support catch handler flags");
+
+        // If this is a catch-all, register that and abort.
+        if (handler.isCatchAll()) {
+          assert(!hasCatchAll);
+          hasCatchAll = true;
+          goto done;
+        }
+
+        cgm.errorNYI("emitLandingPad: non catch-all");
+        return {};
+      }
+
+      goto done;
+    }
+
+  done:
+    if (hasCatchAll) {
+      handlerAttrs.push_back(cir::CatchAllAttr::get(&getMLIRContext()));
+    } else {
+      cgm.errorNYI("emitLandingPad: non catch-all");
+      return {};
+    }
+
+    // Add final array of clauses into TryOp.
+    tryOp.setHandlerTypesAttr(
+        mlir::ArrayAttr::get(&getMLIRContext(), handlerAttrs));
+  }
+
+  // In traditional LLVM codegen. this tells the backend how to generate the
+  // landing pad by generating a branch to the dispatch block. In CIR,
+  // getEHDispatchBlock is used to populate blocks for later filing during
+  // cleanup handling.
+  (void)getEHDispatchBlock(ehStack.getInnermostEHScope(), tryOp);
+
+  return tryOp;
+}
+
+// Differently from LLVM traditional codegen, there are no dispatch blocks
+// to look at given cir.try_call does not jump to blocks like invoke does.
+// However, we keep this around since other parts of CIRGen use
+// getCachedEHDispatchBlock to infer state.
+mlir::Block *
+CIRGenFunction::getEHDispatchBlock(EHScopeStack::stable_iterator scope,
```

Comment:
Does this do anything other than set `mayThrow`? Will it ever do anything else?

Reply:
For now just setting mayThrow, but later I think will be used for emitting Cleanup or unwind regions